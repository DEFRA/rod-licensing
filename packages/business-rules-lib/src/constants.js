/**
 * The maximum days in advance of the purchase date that the licence may start
 * @type {number}
 */
export const ADVANCED_PURCHASE_MAX_DAYS = 30

/**
 * The  maximum number of permissions in a (multi-buy) transaction
 * @type {number}
 */
export const MAX_PERMISSIONS_PER_TRANSACTION = 50

/**
 * Allowed status values for the payment journal
 */
export const PAYMENT_JOURNAL_STATUS_CODES = {
  InProgress: 'In Progress',
  Cancelled: 'Cancelled',
  Failed: 'Failed',
  Expired: 'Expired',
  Completed: 'Completed'
}

/**
 * Payment status returned by the GOV.UK pay API
 */
export const PAYMENT_STATUS = {
  Success: 'success',
  Failure: 'failed'
}

/**
 * Failure states returned by the GOV.UK pay API
 * https://docs.payments.service.gov.uk/api_reference/#errors-caused-by-payment-statuses
 */
export const GOVUK_PAY_ERROR_STATUS_CODES = {
  REJECTED: 'P0010',
  EXPIRED: 'P0020',
  USER_CANCELLED: 'P0030',
  ERROR: 'P0050',
  NOT_FOUND: 'P0200'
}

/**
 * Constants used in finalization
 * @type {{telesales: string, govPay: string}}
 */
export const TRANSACTION_SOURCE = { govPay: 'Gov Pay', telesales: 'Telesales' }
export const PAYMENT_TYPE = { debit: 'Debit card' }

/**
 * Data sources associated with POCL transactions
 */
export const POCL_DATA_SOURCE = 'Post Office Sales'
export const DDE_DATA_SOURCE = 'DDE File'
export const POCL_TRANSACTION_SOURCES = [POCL_DATA_SOURCE, DDE_DATA_SOURCE]

/**
 * Describes the lifetime of the renewal link.
 */
export const RENEW_BEFORE_DAYS = 60 // The number of days before licence expiry the link is active
export const RENEW_AFTER_DAYS = 60 // The number of days after licence expiry the link is active

/**
 * The number of minutes after payment that a licence will start
 * @type {number}
 */
export const START_AFTER_PAYMENT_MINUTES = 30

/**
 * Timezone of the service
 */
export const SERVICE_LOCAL_TIME = 'Europe/London'

/**
 * Date for switching paper fulfilment provider
 */
export const FULFILMENT_SWITCHOVER_DATE = '2024-05-30T23:00:00.000Z'
