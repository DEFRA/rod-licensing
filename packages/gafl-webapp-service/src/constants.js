/**
 * System constants and defaults
 */
export const ADDRESS_LOOKUP_SERVICE = { lang: 'EN', dataset: 'DPA' }
export const ADDRESS_LOOKUP_TIMEOUT_MS_DEFAULT = 10000
export const ATTRIBUTION_REDIRECT_DEFAULT = '/'
export const SESSION_TTL_MS_DEFAULT = 3 * 60 * 60 * 1000
export const PORT_DEFAULT = 3000
export const REDIS_PORT_DEFAULT = 6379
export const PAGE_STATE = { completed: true, error: false }
export const SESSION_COOKIE_NAME_DEFAULT = 'sid'
export const CSRF_TOKEN_COOKIE_NAME_DEFAULT = 'rlsctkn'
export const FEEDBACK_URI_DEFAULT = '#'
export const CHANNEL_DEFAULT = 'websales'
export const SERVICE_PAGE_DEFAULT = 'https://www.gov.uk/fishing-licences/buy-a-fishing-licence'

export const COMPLETION_STATUS = {
  agreed: 'agreed',
  posted: 'posted',
  paymentCreated: 'payment-created',
  paymentCancelled: 'payment-cancelled',
  paymentFailed: 'payment-failed',
  paymentCompleted: 'payment-completed',
  finalised: 'finalised',
  completed: 'completed'
}

export const GOVPAYFAIL = {
  prePaymentRetry: { step: 'pre-payment' },
  postPaymentRetry: { step: 'post-payment' }
}

export const UTM = {
  CAMPAIGN: 'utm_campaign',
  MEDIUM: 'utm_medium',
  SOURCE: 'utm_source',
  TERM: 'utm_term',
  CONTENT: 'utm_content'
}

export const dateFormats = ['YYYY-MM-DD', 'YY-MM-DD', 'YYYY-M-DD', 'YY-M-DD', 'YYYY-MM-D', 'YY-MM-D', 'YYYY-M-D', 'YY-M-D']

export const CommonResults = {
  SUMMARY: 'summary',
  OK: 'ok',
  NO: 'no',
  YES: 'yes'
}

export const allowsPhysicalLicence = {
  YES: 'yes',
  NO: 'no'
}

// If the user has seen the summary page these are set in the status
export const CONTACT_SUMMARY_SEEN = 'contact-summary'
export const LICENCE_SUMMARY_SEEN = 'licence-summary'

// These cookies are used by the load balancer
export const ALB_COOKIE_NAME = 'AWSALBTG'
export const ALBCORS_COOKIE_NAME = 'AWSALBTGCORS'
