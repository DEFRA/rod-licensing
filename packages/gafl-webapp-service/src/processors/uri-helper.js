export const addLanguageCodeToUri = (request, uri) => {
  const path = uri || request.path
  console.log('decorating', path)
  const appendWith = /.*\?.*/.test(path) ? '&' : '?'
  console.log('appendWith', appendWith)
  return `${path}${/\?.*lang=cy.*$/.test(request.url.search) ? `${appendWith}lang=cy` : ''}`
}
