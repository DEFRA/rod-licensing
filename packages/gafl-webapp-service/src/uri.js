/*
 * Page locations, templates
 */
export const LICENCE_LENGTH = { uri: '/buy/licence-length', page: 'licence-length' }
export const LICENCE_TYPE = { uri: '/buy/licence-type', page: 'licence-type' }
export const LICENCE_TO_START = { uri: '/buy/start-kind', page: 'licence-to-start' }
export const LICENCE_START_TIME = { uri: '/buy/start-time', page: 'licence-start-time' }

export const NO_LICENCE_REQUIRED = { uri: '/buy/no-licence-required', page: 'no-licence-required' }
export const DISABILITY_CONCESSION = { uri: '/buy/disability-concession', page: 'disability-concession' }
export const DATE_OF_BIRTH = { uri: '/buy/date-of-birth', page: 'date-of-birth' }

export const NAME = { uri: '/buy/name', page: 'name' }
export const ADDRESS_LOOKUP = { uri: '/buy/find-address', page: 'address-lookup' }
export const ADDRESS_SELECT = { uri: '/buy/select-address', page: 'address-select' }
export const ADDRESS_ENTRY = { uri: '/buy/address', page: 'address-entry' }
export const CONTACT = { uri: '/buy/contact', page: 'contact' }
export const NEWSLETTER = { uri: '/buy/newsletter', page: 'newsletter' }

export const LICENCE_FULFILMENT = { uri: '/buy/fulfilment', page: 'licence-fulfilment' }
export const LICENCE_CONFIRMATION_METHOD = { uri: '/buy/confirmation-method', page: 'licence-confirmation-method' }

export const CONTACT_SUMMARY = { uri: '/buy/contact-summary', page: 'contact-summary' }
export const LICENCE_SUMMARY = { uri: '/buy/licence-summary', page: 'licence-summary' }

export const PAYMENT_CANCELLED = { uri: '/buy/payment-cancelled', page: 'payment-cancelled' }
export const PAYMENT_FAILED = { uri: '/buy/payment-failed', page: 'payment-failed' }

export const TERMS_AND_CONDITIONS = { uri: '/buy/terms-conditions', page: 'terms-and-conditions' }
export const ORDER_COMPLETE = { uri: '/buy/order-complete', page: 'order-complete' }
export const LICENCE_INFORMATION = { uri: '/buy/licence-information', page: 'licence-information' }
export const ATTRIBUTION = { uri: '/buy/attribution' }

export const OIDC_SIGNIN = { uri: '/oidc/signin' }
export const OIDC_ROLE_REQUIRED = { uri: '/oidc/role-required', page: 'role-required' }
export const OIDC_ACCOUNT_DISABLED = { uri: '/oidc/account-disabled', page: 'account-disabled' }

/**
 * Renewals pages
 * @type {{uri: string}}
 */
export const RENEWAL_PUBLIC = { uri: '/renew/{referenceNumber?}' }
export const IDENTIFY = { uri: '/buy/renew/identify', page: 'identify' }
export const RENEWAL_INACTIVE = { uri: '/buy/renew/inactive', page: 'renewal-inactive' }
export const AUTHENTICATE = { uri: '/buy/renew/authenticate' }
export const RENEWAL_START_DATE = { uri: '/buy/renew/renewal-start-date', page: 'renewal-start-date' }
export const RENEWAL_START_VALIDATE = { uri: '/buy/renew/renewal-start-date/validate' }

export const CONTROLLER = { uri: '/buy' }
export const NEW_TRANSACTION = { uri: '/buy/new' }
export const ADD_PERMISSION = { uri: '/buy/add' }
export const AGREED = { uri: '/buy/agreed' }

export const CLIENT_ERROR = { uri: '/buy/client-error', page: 'client-error' }
export const SERVER_ERROR = { uri: '/buy/server-error', page: 'server-error' }

/**
 * These are informational static pages
 */
export const COOKIES = { uri: '/guidance/cookies', page: 'cookies' }
export const ACCESSIBILITY_STATEMENT = { uri: '/guidance/accessibility-statement', page: 'accessibility-statement' }
export const PRIVACY_POLICY = { uri: '/guidance/privacy-policy', page: 'privacy-policy' }
export const REFUND_POLICY = { uri: '/guidance/refund-policy', page: 'refund-policy' }
export const OS_TERMS = { uri: '/guidance/os-terms', page: 'os-terms' }

/**
 * These are inserted at runtime by the test framework but the session manager needs to know about them
 */
export const TEST_STATUS = { uri: '/buy/status' }
export const TEST_TRANSACTION = { uri: '/buy/transaction' }
export const GET_PRICING_TYPES = { uri: '/buy/get-pricing/types' }
export const GET_PRICING_LENGTHS = { uri: '/buy/get-pricing/lengths' }

export const FRESHWATER_FISING_RULES = { uri: 'https://www.gov.uk/freshwater-rod-fishing-rules' }
export const LOCAL_BYELAWS = { uri: 'https://www.gov.uk/government/collections/local-fishing-byelaws' }
