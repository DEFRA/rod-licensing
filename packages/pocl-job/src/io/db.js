import config from '../config.js'
import { docClient } from '../../../connectors-lib/src/aws.js'
import { GetCommand, UpdateCommand } from '@aws-sdk/lib-dynamodb'

/**
 * Update the POCL file staging table to add or update the entry for the provided filename
 *
 * @param {string} filename the primary key of the record in the file staging table to add/update
 * @param {Object} entries to update the table with
 * @returns {Promise<void>}
 */
export const updateFileStagingTable = async ({ filename, ...entries }) => {
  const updateExpression = docClient.createUpdateExpression({
    expires: Math.floor(Date.now() / 1000) + config.db.stagingTtlDelta,
    ...entries
  })
  await docClient.send(
    new UpdateCommand({
      TableName: config.db.fileStagingTable,
      Key: { filename },
      ...updateExpression
    })
  )
}

/**
 * Retrieve all file records for the specified stages.  If stages are not provided then all records are returned
 *
 * @param {string} stages the stage names to filter the result-set on. Omit to retrieve all records regardless of the stage.
 * @returns {Promise<[DocumentClient.AttributeMap]>}
 */
export const getFileRecords = async (...stages) => {
  const stageValues = stages.reduce((acc, s, i) => ({ ...acc, [`:stage${i}`]: s }), {})
  return docClient.scanAllPromise({
    TableName: config.db.fileStagingTable,
    ...(stages.length && { FilterExpression: `stage IN (${Object.keys(stageValues)})` }),
    ExpressionAttributeValues: stageValues,
    ConsistentRead: true
  })
}

/**
 * Retrieve an individual file record by the specified filename
 *
 * @param filename the name of the POCL file to retrieve a record for
 * @returns {DocumentClient.AttributeMap}
 */
export const getFileRecord = async filename => {
  const result = await docClient.send(
    new GetCommand({
      TableName: config.db.fileStagingTable,
      Key: { filename },
      ConsistentRead: true
    })
  )
  return result.Item
}

/**
 * Update the POCL record staging table to add entries for each of the provided records
 *
 * @param {string} filename the filename of a POCL file to which the records relate
 * @param {Array<Object>} records to update the table with
 * @returns {Promise<void>}
 */
export const updateRecordStagingTable = async (filename, records) => {
  if (records.length) {
    const params = {
      RequestItems: {
        [config.db.recordStagingTable]: records.map(record => ({
          PutRequest: { Item: { filename, expires: Math.floor(Date.now() / 1000) + config.db.stagingTtlDelta, ...record } }
        }))
      }
    }
    await docClient.batchWriteAllPromise(params)
  }
}

/**
 * Retrieve processed records from the record staging table for the provided filename and stage filters
 *
 * @param {string} filename the filename of a POCL file for which the records should be retrieved
 * @param {string} stages the stage names to filter the result-set on. Omit to retrieve all records regardless of the stage.
 * @returns {Promise<[DocumentClient.AttributeMap]>}
 */
export const getProcessedRecords = async (filename, ...stages) => {
  const stageValues = stages.reduce((acc, s, i) => ({ ...acc, [`:stage${i}`]: s }), {})
  return docClient.queryAllPromise({
    TableName: config.db.recordStagingTable,
    KeyConditionExpression: 'filename = :filename',
    // ...(stages.length && { FilterExpression: `stage IN (${Object.keys(stageValues).join(',')})` }),
    ...(stages.length && { FilterExpression: `stage IN (${Object.keys(stageValues)})` }),
    ExpressionAttributeValues: { ':filename': filename, ...stageValues },
    ConsistentRead: true
  })
}
