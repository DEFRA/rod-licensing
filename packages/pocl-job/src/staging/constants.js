export const MAX_CREATE_TRANSACTION_BATCH_SIZE = 25
export const MAX_FINALISE_TRANSACTION_BATCH_SIZE = 10

export const POST_OFFICE_DATASOURCE = 'Post Office Sales'
export const DIRECT_DEBIT_DATASOURCE = 'DDE File'
export const DIRECT_DEBIT_PAYMENTSOURCE = 'Direct Debit'
export const POSTAL_ORDER_DATASOURCE = 'Postal Order Sales'
export const POSTAL_ORDER_PAYMENTSOURCE = 'Postal Order'

export const FILE_STAGE = {
  Pending: 'Received and Pending',
  Staging: 'Staging Transactions',
  Finalising: 'Finalising Transactions',
  Completed: 'Completed'
}

export const DYNAMICS_IMPORT_STAGE = {
  Pending: 'Received and Pending',
  InProgress: 'In Progress',
  Processed: 'Processed',
  ProcessedWithWarnings: 'Processed with Warnings',
  Failed: 'Failed',
  UnderReview: 'Under Review',
  Closed: 'Closed',
  isAlreadyProcessed: status => status !== DYNAMICS_IMPORT_STAGE.Pending && status !== DYNAMICS_IMPORT_STAGE.InProgress
}

export const RECORD_STAGE = {
  TransactionCreated: 'Transaction Created',
  TransactionCreationFailed: 'Transaction Creation Failed',
  TransactionFinalised: 'Transaction Finalised',
  TransactionFinalisationFailed: 'Transaction Finalisation Failed'
}
