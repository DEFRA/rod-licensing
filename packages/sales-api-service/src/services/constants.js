const CONCESSION = {
  SENIOR: 'Senior',
  JUNIOR: 'Junior',
  DISABLED: 'Disabled'
}

const CONCESSION_PROOF = {
  none: 'No Proof',
  blueBadge: 'Blue Badge',
  NI: 'National Insurance Number'
}

export { CONCESSION, CONCESSION_PROOF }
