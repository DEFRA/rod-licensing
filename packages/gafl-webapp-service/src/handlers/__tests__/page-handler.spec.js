import db from 'debug'
import pageHandler from '../page-handler.js'
import journeyDefinition from '../../routes/journey-definition.js'
import { addLanguageCodeToUri } from '../../processors/uri-helper.js'
import GetDataRedirect from '../get-data-redirect.js'
import { ANALYTICS } from '../../constants.js'
import { AGREED, IDENTIFY, LICENCE_DETAILS, LICENCE_FOR, ORDER_COMPLETE, PAYMENT_CANCELLED, PAYMENT_FAILED } from '../../uri.js'

jest.mock('debug', () => jest.fn(() => jest.fn()))
jest.mock('../../routes/journey-definition.js', () => [])
jest.mock('../../processors/uri-helper.js')

jest.mock('../../constants', () => ({
  ANALYTICS: {
    selected: 'selected',
    acceptTracking: 'accepted-tracking',
    seenMessage: 'seen-message'
  },
  PAGE_STATE: { completed: true, error: false }
}))

jest.mock('../../uri.js', () => ({
  AGREED: { uri: '/buy/agreed/page' },
  LICENCE_DETAILS: { uri: '/buy/licence/details' },
  ORDER_COMPLETE: { uri: '/buy/order/complete/page' },
  PAYMENT_CANCELLED: { uri: '/buy/payment/cancelled/page' },
  PAYMENT_FAILED: { uri: '/buy/payment/failed/page' },
  PROCESS_ANALYTICS_PREFERENCES: { uri: '/buy/process/analytics/preferences/page' },
  LICENCE_FOR: { uri: '/buy/licence-for' },
  IDENTIFY: { uri: '/buy/renew/identify' }
}))

describe('The page handler function', () => {
  let fakeDebug

  beforeAll(() => {
    fakeDebug = db.mock.results[0].value
  })
  beforeEach(() => {
    jest.resetAllMocks()
    journeyDefinition.length = 0
  })

  it('the get method re-throws any exceptions which are not transaction errors ', async () => {
    const request = getMockRequest()

    const getData = async () => {
      throw new Error('Random exception')
    }

    try {
      await pageHandler(null, null, null, getData).get(request)
    } catch (err) {
      expect(err.message).toBe('Random exception')
    }
  })

  it('the error method re-throws any exceptions which are not transaction errors ', async () => {
    const request = getMockRequest(() => {
      throw new Error('Random exception')
    })
    try {
      await pageHandler().error(request, null, { details: [] })
    } catch (err) {
      expect(err.message).toBe('Random exception')
    }
  })

  it('logs the cache if getCurrentPermission throws an error', async () => {
    const request = {
      cache: () => ({
        helpers: {
          page: {
            getCurrentPermission: () => {
              throw new Error('Random exception')
            },
            get: () => ({})
          },
          status: {
            get: () => ({})
          },
          transaction: {
            get: () => ({})
          },
          analytics: {
            get: () => ({})
          }
        }
      })
    }
    await expect(pageHandler().get(request)).rejects.toThrow('Random exception')
    expect(fakeDebug).toHaveBeenCalledWith(expect.stringContaining('Page cache'))
    expect(fakeDebug).toHaveBeenCalledWith(expect.stringContaining('Status cache'))
    expect(fakeDebug).toHaveBeenCalledWith(expect.stringContaining('Transaction cache'))
    expect(fakeDebug).toHaveBeenCalledWith(expect.stringContaining('Analytics cache'))
  })

  it.each([['/previous/page'], ['/last/page']])('get calls addLanguageCodeToUri with request and backLink', async previousPage => {
    const backLink = () => previousPage
    journeyDefinition.push({ current: { page: 'view' }, backLink })
    const { get } = pageHandler(null, 'view', '/next/page')
    const request = getMockRequest(undefined, '/this/page')
    const toolkit = getMockToolkit()

    await get(request, toolkit)

    expect(addLanguageCodeToUri).toHaveBeenCalledWith(request, backLink())
  })

  it.each([[() => '/last/page'], ['previous/page']])(
    'pageData.backRef from get uses value generated by addLanguageCodeToUri',
    async backLink => {
      journeyDefinition.push({ current: { page: 'view' }, backLink })
      const { get } = pageHandler(null, 'view', '/next/page')
      const returnValue = Symbol('/previous/page')
      addLanguageCodeToUri.mockReturnValueOnce(returnValue)
      const toolkit = getMockToolkit()

      await get(getMockRequest(), toolkit)

      expect(toolkit.view).toHaveBeenCalledWith(
        expect.any(String),
        expect.objectContaining({
          backRef: returnValue
        })
      )
    }
  )

  it('pageData.backRef is null if backLink is null', async () => {
    const backLink = () => {}
    journeyDefinition.push({ current: { page: 'view' }, backLink })
    const { get } = pageHandler(null, 'view', '/next/page')
    const returnValue = Symbol('/previous/page')
    addLanguageCodeToUri.mockReturnValueOnce(returnValue)
    const toolkit = getMockToolkit()

    await get(getMockRequest(), toolkit)

    expect(toolkit.view).toHaveBeenCalledWith(
      expect.any(String),
      expect.objectContaining({
        backRef: null
      })
    )
  })

  it.each([['/go/somewhere'], ['/go/somewhere/else']])(
    'GetDataRedirect being thrown will pass url to be decorated to addLanguageCodeToUri',
    async redirectUri => {
      journeyDefinition.push({ current: { page: 'view' }, backLink: '/previous/page' })
      const getData = () => {
        throw new GetDataRedirect(redirectUri)
      }
      const { get } = pageHandler(null, 'view', '/next/page', getData)
      const toolkit = getMockToolkit()
      const request = getMockRequest()

      await get(request, toolkit)

      expect(addLanguageCodeToUri).toHaveBeenCalledWith(request, redirectUri)
    }
  )

  it('GetDataRedirect being thrown will use addLanguageCodeToUri to decorate the redirect target', async () => {
    journeyDefinition.push({ current: { page: 'view' }, backLink: '/previous/page' })
    const getData = () => {
      throw new GetDataRedirect('/go/somewhere/else')
    }
    const { get } = pageHandler(null, 'view', '/next/page', getData)
    const returnValue = Symbol('/previous/page')
    addLanguageCodeToUri.mockReturnValueOnce(returnValue)
    const toolkit = getMockToolkit()

    await get(getMockRequest(), toolkit)

    expect(toolkit.redirect).toHaveBeenCalledWith(returnValue)
  })

  it('error calls addLanguageCodeToUri with request', async () => {
    const { error } = pageHandler('', 'view')
    const request = getMockRequest()
    await error(request, getMockToolkit(), { details: [] })
    expect(addLanguageCodeToUri).toHaveBeenCalledWith(request)
  })

  it.each([['/route/one'], ['/route/sixty-six']])('error redirects to uri decorated by addLanguageCodeToUri', async url => {
    const { error } = pageHandler('', 'view')
    const toolkit = getMockToolkit()
    addLanguageCodeToUri.mockReturnValueOnce(`Redirect to url ${url}`)
    await error(getMockRequest(undefined, url), toolkit, { details: [] })
    expect(toolkit.redirect).toHaveBeenCalledWith(`Redirect to url ${url}`)
  })

  it('sets the value of pageData with displayAnalytics true', async () => {
    addLanguageCodeToUri.mockReturnValueOnce('/buy/process-analytics-preferences')
    const { get } = pageHandler('', 'view', '/next/page')
    const toolkit = getMockToolkit()
    await get(getMockRequest(), toolkit)
    expect(toolkit.view).toMatchSnapshot()
  })

  it('sets the value of pageData with displayAnalytics false', async () => {
    addLanguageCodeToUri.mockReturnValueOnce('/buy/process-analytics-preferences')
    const { get } = pageHandler('', 'view', '/next/page')
    const toolkit = getMockToolkit()
    await get(getMockRequest(null, '/we/are/here'), toolkit)
    expect(toolkit.view).toMatchSnapshot()
  })

  it.each([
    ['payment cancelled', PAYMENT_CANCELLED.uri],
    ['payment failed', PAYMENT_FAILED.uri],
    ['agreed', AGREED.uri],
    ['order complete', ORDER_COMPLETE.uri],
    ['licence details', LICENCE_DETAILS.uri]
  ])('hides the analytics banner for %s page', async (_pageLabel, pageUri) => {
    const { get } = pageHandler('', 'view', '/next/page')
    const toolkit = getMockToolkit()
    const mockRequest = getMockRequest(null, pageUri)
    await get(mockRequest, toolkit)
    const pageData = toolkit.view.mock.calls[0][1]
    expect(pageData.displayAnalytics).toBeFalsy()
  })

  it('sets analytics values to default values if analytics key is not set', async () => {
    const { get } = pageHandler('', 'view', '/next/page')
    const toolkit = getMockToolkit()
    const mockRequest = getMockRequest(null, '/current/page', false)
    await get(mockRequest, toolkit)
    const pageData = toolkit.view.mock.calls[0][1]
    expect(pageData).toEqual(
      expect.objectContaining({
        analyticsMessageDisplayed: false,
        analyticsSelected: false,
        acceptedTracking: false
      })
    )
  })

  it.each([[PAYMENT_CANCELLED.uri], [PAYMENT_FAILED.uri], [AGREED.uri], [ORDER_COMPLETE.uri], [LICENCE_DETAILS.uri]])(
    'sets journeyBeginning to false if not on licence_for or identify page',
    async pageUri => {
      const { get } = pageHandler('', 'view', '/next/page')
      const toolkit = getMockToolkit()
      const mockRequest = getMockRequest(null, pageUri)
      await get(mockRequest, toolkit)
      const pageData = toolkit.view.mock.calls[0][1]
      expect(pageData).toEqual(
        expect.objectContaining({
          journeyBeginning: false
        })
      )
    }
  )

  // it.only('sets journeyBeginning to true if on licence_for or identify page', async () => {
  //   const { get } = pageHandler('', 'view', '/next/page')
  //   const toolkit = getMockToolkit()
  //   const mockRequest = getMockRequest(null, IDENTIFY.uri)
  //   await get(mockRequest, toolkit)
  //   const pageData = toolkit.view.mock.calls[0][1]
  //   expect(pageData).toEqual(
  //     expect.objectContaining({
  //       journeyBeginning: true
  //     })
  //   )
  // })

  it.each([[IDENTIFY.uri], [LICENCE_FOR.uri]])('sets journeyBeginning to true if on licence_for or identify page', async pageUri => {
    const { get } = pageHandler('', 'view', '/next/page')
    const toolkit = getMockToolkit()
    const mockRequest = getMockRequest(null, pageUri)
    await get(mockRequest, toolkit)
    const pageData = toolkit.view.mock.calls[0][1]
    expect(pageData).toEqual(
      expect.objectContaining({
        journeyBeginning: true
      })
    )
  })
})

const getMockRequest = (setCurrentPermission = () => {}, path = '/buy/we/are/here', includeAnalytics = true) => ({
  cache: () => ({
    helpers: {
      page: {
        getCurrentPermission: () => ({}),
        setCurrentPermission
      },
      status: {
        getCurrentPermission: () => ({}),
        setCurrentPermission: () => {}
      },
      transaction: {
        getCurrentPermission: () => {}
      },
      analytics: {
        get: () =>
          // prettier-ignore
          includeAnalytics
            ? {
                [ANALYTICS.selected]: 'selected',
                [ANALYTICS.acceptTracking]: 'accepted-tracking',
                [ANALYTICS.seenMessage]: 'seen-message'
              }
            : undefined
      }
    }
  }),
  i18n: {
    getCatalog: () => {},
    getLocales: () => [],
    getLocale: () => ''
  },
  path,
  url: {
    search: ''
  }
})

const getMockToolkit = () => ({
  redirect: jest.fn(() => ({ takeover: () => {} })),
  view: jest.fn()
})
