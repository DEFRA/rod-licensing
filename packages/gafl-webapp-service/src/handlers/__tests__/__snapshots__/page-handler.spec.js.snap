// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The page handler function sets the value of pageData 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "view",
      Object {
        "acceptedTracking": "accepted-tracking",
        "altLang": Array [],
        "analyticsMessageDisplayed": "seen-message",
        "analyticsSelected": "selected",
        "backRef": null,
        "displayAnalytics": true,
        "mssgs": undefined,
        "uri": Object {
          "analyticsFormAction": "/buy/process-analytics-preferences",
        },
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;
